src/utils/JobLogUtils.py

```diff
@@ -15,45 +16,112 @@ def TryParseCursor(cursor):
+            return ({}, None)
+elif config.get("logging") == 'elasticsearch':
+    from elasticsearch import Elasticsearch
+
+    def GetJobLog(jobId, cursor=None, size=None):
+        try:
+            elasticsearch = Elasticsearch(config['elasticsearch'])
+
+            request_json = {
+                "query": {
+                    "match_phrase": {
+                        "kubernetes.labels.jobId": jobId
                     }
                 },
-            ],
-            "_source": [
-                "docker.container_id", "kubernetes.pod_name", "stream", "log"
-            ]
-        }
-        if cursor is not None:
-            search_after = TryParseCursor(cursor)
-            if search_after is not None:
-                request_json['search_after'] = search_after
-        if size is not None:
-            request_json['size'] = size
-
-        response_json = elasticsearch.search(index="logstash-*",
-                                             body=request_json)
-        documents = response_json["hits"]["hits"]
-
-        next_cursor = None
-        if len(documents) > 0:
-            next_cursor = '.'.join(str(i) for i in documents[-1]["sort"])
-
-        return (documents, next_cursor)
-    except Exception:
-        logger.exception("Request elasticsearch failed")
-        return ({}, None)
+                "sort": [
+                    "@timestamp",
+                    {
+                        "time_nsec": {
+                            "unmapped_type": "long",
+                            "missing": 0
+                        }
+                    },
+                ],
+                "_source": [
+                    "docker.container_id", "kubernetes.pod_name", "stream",
+                    "log"
+                ]
+            }
+            if cursor is not None:
+                search_after = TryParseCursor(cursor)
+                if search_after is not None:
+                    request_json['search_after'] = search_after
+            if size is not None:
+                request_json['size'] = size
+
+            response_json = elasticsearch.search(index="logstash-*",
+                                                 body=request_json)
+            documents = response_json["hits"]["hits"]
+
+            pod_logs = dict()
+            for pod_name, pod_documents in groupby(
+                    documents, lambda document: document["_source"][
+                        "kubernetes"]["pod_name"]):
+                pod_logs[pod_name] = ''.join(
+                    pod_document["_source"]["log"] for pod_document in pod_documents)
+
+            if len(documents) > 0:
+                cursor = '.'.join(str(i) for i in documents[-1]["sort"])
+            else:
+                cursor = None
+
+            return (pod_logs, cursor)
+        except Exception:
+            logger.exception("Request elasticsearch failed")
+            return ({}, None)
+else:
```

Comment:
Maybe add a logging in each branch to show which backend is chosen to make debug faster?

Reply:
Added

Reply:
No, I meant add log outside of function, so the log only shown once. e.g.

```
if config.get("logging") == 'azureBlob':
    logger.info('Requesting Azure Blob to get log of the job %s', jobId)
    from azure.storage.blob import AppendBlobService
...
```

Reply:
Got it

Reply:
I see